xds/internal/balancer/edsbalancer/balancergroup.go

```diff
@@ -167,50 +287,41 @@ func (bg *balancerGroup) changeWeight(id internal.Locality, newWeight uint32) {
 		return
 	}
 	pState.weight = newWeight
-	// Update state and picker to reflect the changes.
-	bg.cc.UpdateBalancerState(buildPickerAndState(bg.idToPickerState))
+	if bg.pickerStarted {
+		// Update state and picker to reflect the changes.
+		bg.cc.UpdateBalancerState(buildPickerAndState(bg.idToPickerState))
+	}
 }
 
 // Following are actions from the parent grpc.ClientConn, forward to sub-balancers.
 
 // SubConn state change: find the corresponding balancer and then forward.
 func (bg *balancerGroup) handleSubConnStateChange(sc balancer.SubConn, state connectivity.State) {
 	grpclog.Infof("balancer group: handle subconn state change: %p, %v", sc, state)
-	bg.mu.Lock()
-	var b balancer.Balancer
+	bg.incomingMu.Lock()
 	if id, ok := bg.scToID[sc]; ok {
 		if state == connectivity.Shutdown {
 			// Only delete sc from the map when state changed to Shutdown.
 			delete(bg.scToID, sc)
 		}
-		b = bg.idToBalancer[id]
-	}
-	bg.mu.Unlock()
-	if b == nil {
-		grpclog.Infof("balancer group: balancer not found for sc state change")
-		return
-	}
-	if ub, ok := b.(balancer.V2Balancer); ok {
-		ub.UpdateSubConnState(sc, balancer.SubConnState{ConnectivityState: state})
-	} else {
-		b.HandleSubConnStateChange(sc, state)
+		bg.outgoingMu.Lock()
+		config, ok := bg.idToBalancerConfig[id]
+		if ok {
+			config.handleSubConnStateChange(sc, state)
+		}
+		bg.outgoingMu.Unlock()
 	}
+	bg.incomingMu.Unlock()
 }
 
 // Address change: forward to balancer.
 func (bg *balancerGroup) handleResolvedAddrs(id internal.Locality, addrs []resolver.Address) {
-	bg.mu.Lock()
-	b, ok := bg.idToBalancer[id]
-	bg.mu.Unlock()
-	if !ok {
-		grpclog.Infof("balancer group: balancer with id %q not found", id)
-		return
-	}
-	if ub, ok := b.(balancer.V2Balancer); ok {
-		ub.UpdateClientConnState(balancer.ClientConnState{ResolverState: resolver.State{Addresses: addrs}})
-	} else {
-		b.HandleResolvedAddrs(addrs, nil)
+	bg.outgoingMu.Lock()
+	config, ok := bg.idToBalancerConfig[id]
```

Comment:
Same thing here, could be combined in the same line.
And should we handle the case where the entry is not found in the map.

Reply:
Done.

Not found should only happen when there's a race between remove balancer and this function. We should ignore the address in this case.